### PR TITLE
Bundle tool to create a self-containing Arcs feature bundle

### PR DESCRIPTION
**Pull request by piotrswigon**

### CODE REVIEW

File: src/tools/tests/bundle-test.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+          {filePath: 'src/tools/tests/test-data/a/a.js', bundlePath: 'a/a.js', entryPoint: false},
+          {filePath: 'src/tools/tests/test-data/a/a.manifest', bundlePath: 'a/a.manifest', entryPoint: true},
+          {filePath: 'src/tools/tests/test-data/b/b.json', bundlePath: 'b/b.json', entryPoint: false},
+          {filePath: 'src/tools/tests/test-data/b/b.manifest', bundlePath: 'b/b.manifest', entryPoint: true},
+        ]
+    );
+  });
+  it('bundles Products demo', done => {
+    bundle(['src/runtime/test/artifacts/Products/Products.recipes'], 'test-output/bundle/products.zip', false).then(() => {
+      fs.readFile('test-output/bundle/products.zip', (err, data) => {
```
Comment (mykmartin) [inline src/tools/tests/bundle-test.ts]: fs.readFileSync?

File: src/tools/bundle-cli.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ * @license
+ * Copyright (c) 2019 Google Inc. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import minimist from 'minimist';
```
Comment (lindner) [inline src/tools/bundle-cli.ts]: suggest replacing minimist with something else.  It's obsolete and unmaintained.

File: src/tools/bundle.ts [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+/**
+ * @param entryPoints array of paths to Arc manifests to bundle, e.g. ['./feature/awesome.recipes']
+ * @param bundleName path to the output bundle, e.g. './awesome.zip'
+ * @param verbose whether to print bundled files to stdout
+ */
+export async function bundle(entryPoints: string[], bundleName: string, verbose: boolean) {
+  const listing = await bundleListing(...entryPoints);
+
+  if (verbose) {
+    console.log(`Bundled files:`);
```
[the remark this answers is not in the document]
Comment (piotrswigon) [reply src/tools/bundle.ts]: Done

File: src/tools/bundle.ts [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        .map(f => `entry-point: ${f.bundlePath}\n`)
+        .join(''));
+    archive.generateNodeStream({streamFiles:true})
+        .pipe(fs.createWriteStream(bundleName))
+        .on('finish', () => resolve());
+  });
+}
+
+export async function bundleListing(...entryPoints: string[]): Promise<BundleEntry[]> {
+  const entryManifests = await Promise.all(entryPoints.map(async ep => Manifest.load(ep, new Loader())));
```
[the remark this answers is not in the document]
Comment (piotrswigon) [reply src/tools/bundle.ts]: Done both, thanks!

File: src/tools/bundle.ts [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+  
+  let dependencies: string[] = [];
+  entryManifests.forEach(m => collectDependencies(m, dependencies));
+  dependencies = [...new Set(dependencies)];
+  dependencies.sort();
```
[the remark this answers is not in the document]
Comment (piotrswigon) [reply src/tools/bundle.ts]: Done :)

File: src/tools/bundle.ts [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+  const prefixLengthToSubtract = dirPrefixForSortedPaths(dependencies);
+  const entryManifestsFileNames = entryManifests.map(m => m.fileName);
+
+  return dependencies.map(filePath => ({
+    filePath,
+    bundlePath: filePath.substr(prefixLengthToSubtract),
```
[the remark this answers is not in the document]
Comment (piotrswigon) [reply src/tools/bundle.ts]: I decided that the cleanest way to do this is to reflect the directory structure of the source. You can see this in tests, but essentially I preserve directory names, truncating the common prefix.

So if entry point is: /a/b/c/d/e/g/start.manifest

and imports ../../../h/i/j/import.manifest

then bundle will have:

d/e/g/start.manifest
h/i/j/import.manifest

and a bundle-manifest.mf file containing:

entry-point: d/e/g/start.manifest

I started with some ideas for renaming directories etc, but I decided this is the cleanest solution.

Some day we may add --obfuscate to obfuscate the names of files and directories, or even --minimize to roll everything up into a single manifest file, but let's not over engineer for now :)

I was considering doing a half-pager describing this design, WDYT?

File: src/tools/tests/bundle-test.ts (hunk 1 above)
Comment (piotrswigon) [reply src/tools/tests/bundle-test.ts]: Right! I managed to simplify the async stuff here. I don't remember know why I thought this has to be so complex :/

File: src/tools/bundle.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (mykmartin) [reply src/tools/bundle.ts]: Thanks for the detailed explanation; LGTM.

File: src/tools/bundle-cli.ts (hunk 2 above)
Comment (mykmartin) [reply src/tools/bundle-cli.ts]: sigh is already using minimist, and given the scope (basic args parsing) being unmaintained isn't much of a problem - we don't currently need sophisticated parsing and it's unlikely to break in the foreseeable future.

Comment (mykmartin) [reply src/tools/bundle-cli.ts]: Just to clarify: I'm not against replacing minimist with something newer, but it makes sense for this PR to use something we already depend on (and know about).

Comment (piotrswigon) [reply src/tools/bundle-cli.ts]: I'll leave as is on the basis that that's what we already use in our repo. Let's migrate to a new solution across all usages.

**Pull request merged**